src/signature.c

```diff
@@ -2009,3 +2049,274 @@ const OSSL_DISPATCH p11prov_ecdsa_signature_functions[] = {
+                                       size_t siglen, const unsigned char *tbs,
+                                       size_t tbslen)
+{
+    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
+    CK_RV ret;
+
+    P11PROV_debug("eddsa digest verify (ctx=%p, tbs=%p, tbslen=%zu)", ctx, tbs,
+                  tbslen);
+
+    if (sigctx == NULL) {
+        return RET_OSSL_ERR;
+    }
+
+    ret = p11prov_sig_operate(sigctx, (void *)sig, NULL, siglen, (void *)tbs,
+                              tbslen);
+    if (ret != CKR_OK) {
+        return RET_OSSL_ERR;
+    }
+    return RET_OSSL_OK;
+}
+
+static int p11prov_eddsa_get_ctx_params(void *ctx, OSSL_PARAM *params)
+{
+    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
+    OSSL_PARAM *p;
+    int ret;
+
+    /* todo sig params:
+        OSSL_SIGNATURE_PARAM_ALGORITHM_ID
+     */
+
+    P11PROV_debug("eddsa get ctx params (ctx=%p, params=%p)", ctx, params);
+
+    p = OSSL_PARAM_locate(params, OSSL_SIGNATURE_PARAM_ALGORITHM_ID);
+    if (p) {
+        if (sigctx->mechtype != CKM_EDDSA) {
+            return RET_OSSL_ERR;
+        }
+        CK_ULONG size = p11prov_obj_get_key_bit_size(sigctx->key);
+        switch (size) {
+        case ED25519_BIT_SIZE:
+            ret = OSSL_PARAM_set_octet_string(p, der_ed25519_algorithm_id,
+                                              sizeof(der_ed25519_algorithm_id));
+            break;
+        case ED448_BIT_SIZE:
+            ret = OSSL_PARAM_set_octet_string(p, der_ed448_algorithm_id,
+                                              sizeof(der_ed448_algorithm_id));
+            break;
+        default:
+            ret = RET_OSSL_ERR;
+            break;
+        }
+        if (ret != RET_OSSL_OK) {
+            return ret;
+        }
+    }
+
+    return RET_OSSL_OK;
+}
+
+#ifndef OSSL_SIGNATURE_PARAM_INSTANCE
+#define OSSL_SIGNATURE_PARAM_INSTANCE "instance"
+#endif
+#ifndef OSSL_SIGNATURE_PARAM_CONTEXT_STRING
+#define OSSL_SIGNATURE_PARAM_CONTEXT_STRING "context-string"
+#endif
+static int p11prov_eddsa_set_ctx_params(void *ctx, const OSSL_PARAM params[])
+{
+    P11PROV_SIG_CTX *sigctx = (P11PROV_SIG_CTX *)ctx;
+    const OSSL_PARAM *p;
+    int ret;
+
+    P11PROV_debug("eddsa set ctx params (ctx=%p, params=%p)", sigctx, params);
+
+    if (params == NULL) {
+        return RET_OSSL_OK;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_SIGNATURE_PARAM_INSTANCE);
+    if (p) {
+        const char *instance = NULL;
+        bool matched = false;
+        CK_ULONG size;
+
+        ret = OSSL_PARAM_get_utf8_string_ptr(p, &instance);
+        if (ret != RET_OSSL_OK) {
+            return ret;
+        }
+        P11PROV_debug("Set OSSL_SIGNATURE_PARAM_INSTANCE to %s", instance);
+
+        size = p11prov_obj_get_key_bit_size(sigctx->key);
+        if (size != ED25519_BIT_SIZE && size != ED448_BIT_SIZE) {
+            P11PROV_raise(sigctx->provctx, CKR_KEY_TYPE_INCONSISTENT,
+                          "Invalid EdDSA key size %lu\n", size);
```

Comment:
the log/error messages do not need the trailing newline as it is added by the logging functions already

Reply:
good catch